File: scripts/cleanup_duplicate_docstrings.py
#!/usr/bin/env python3
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Remove duplicate/attribute docstrings from class and module bodies."""

from __future__ import annotations

import ast
import sys
from pathlib import Path


def remove_extra_docstrings(source: str) -> str:
    """Remove standalone string literals that are not the first statement in a class/module body."""
    try:
        tree = ast.parse(source)
    except SyntaxError:
        return source

    lines = source.splitlines(keepends=True)
    removals = []

    for node in ast.walk(tree):
        if not isinstance(node, (ast.Module, ast.ClassDef)):
            continue

        body = node.body
        if not body:
            continue

        # Check if first statement is a docstring
        first_is_docstring = (
            isinstance(body[0], ast.Expr)
            and isinstance(body[0].value, ast.Constant)
            and isinstance(body[0].value.value, str)
        )

        if not first_is_docstring:
            # Remove ALL standalone string literals in the body (except we keep first if it's docstring)
            pass

        for i in range(len(body)):
            stmt = body[i]
            if not isinstance(stmt, ast.Expr):
                continue
            if not isinstance(stmt.value, ast.Constant):
                continue
            if not isinstance(stmt.value.value, str):
                continue

            if i == 0 and first_is_docstring:
                continue  # Keep the official docstring

            # This is an extra string literal - remove it
            # Find the line range for this statement
            start_line = stmt.lineno - 1  # 0-indexed
            end_line = stmt.end_lineno - 1

            # Also remove trailing blank lines after the string literal
            while end_line + 1 < len(lines) and lines[end_line + 1].strip() == "":
                end_line += 1

            removals.append((start_line, end_line))

    if not removals:
        return source

    # Sort and merge removals
    removals.sort(reverse=True)

    for start, end in removals:
        # Remove lines[start:end+1]
        del lines[start : end + 1]

    return "".join(lines)


def process_file(path: Path) -> bool:
    source = path.read_text(encoding="utf-8")
    cleaned = remove_extra_docstrings(source)

    if cleaned == source:
        return False

    # Verify syntax
    try:
        ast.parse(cleaned)
    except SyntaxError as e:
        print(f"  SKIP (would break syntax): {path} — {e}")
        return False

    path.write_text(cleaned, encoding="utf-8")
    print(f"  CLEANED: {path}")
    return True


def main():
    files = sys.argv[1:]
    if not files:
        print("Usage: python cleanup_duplicate_docstrings.py <file1.py> [file2.py] ...")
        sys.exit(1)

    total = 0
    for f in files:
        if process_file(Path(f)):
            total += 1

    print(f"\nTotal files cleaned: {total}")


if __name__ == "__main__":
    main()

File: src/typescript/src/maintenance/cleanupDuplicateDocstrings.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { errorMessage, parseFileArguments, readUtf8Text, writeTextFile, type WriteOptions } from './io.js'
import { analyzePythonSource, applySourceEdits, barePlainStringTokens, lineRemovalRange } from './pythonSource.js'

export interface DuplicateDocstringResult {
  changed: boolean
  removed: number
  text: string
  valid: boolean
}

export interface DuplicateDocstringFileResult extends DuplicateDocstringResult {
  path: string
}

/** Remove non-docstring standalone literals from module and class bodies. */
export function cleanupDuplicateDocstrings(source: string): DuplicateDocstringResult {
  const analysis = analyzePythonSource(source)
  if (!analysis.valid) return { changed: false, removed: 0, text: source, valid: false }

  const removals = analysis.contexts.flatMap(context => {
    if (context.scope.kind !== 'class' && context.scope.kind !== 'module') return []
    if (context.firstInScope || barePlainStringTokens(context.statement) === undefined) return []
    return [lineRemovalRange(source, context.statement.start, context.statement.end)]
  })
  const text = applySourceEdits(source, removals)
  const valid = analyzePythonSource(text).valid
  if (!valid) return { changed: false, removed: 0, text: source, valid: false }
  return { changed: text !== source, removed: removals.length, text, valid: true }
}

/** Transform explicitly named Python files with dry-run and atomic-write controls. */
export async function cleanupDuplicateDocstringFiles(
  paths: readonly string[],
  options: WriteOptions = {},
): Promise<DuplicateDocstringFileResult[]> {
  const results: DuplicateDocstringFileResult[] = []
  for (const path of paths) {
    const source = await readUtf8Text(path)
    const result = cleanupDuplicateDocstrings(source)
    if (result.changed) await writeTextFile(path, result.text, options)
    results.push({ ...result, path })
  }
  return results
}

export async function main(args: readonly string[] = process.argv.slice(2)): Promise<number> {
  try {
    const parsed = parseFileArguments(
      args,
      'Usage: bun cleanupDuplicateDocstrings.ts [--dry-run] [--no-atomic] <file1.py> [file2.py ...]',
    )
    if (parsed === undefined) return 0
    if (parsed.paths.length === 0) throw new Error('At least one Python source path is required.')

    const results = await cleanupDuplicateDocstringFiles(parsed.paths, parsed)
    let total = 0
    for (const result of results) {
      if (!result.valid) {
        console.log(`  SKIP (lexically invalid Python): ${result.path}`)
        continue
      }
      if (!result.changed) continue
      total += 1
      console.log(`  ${parsed.dryRun ? 'WOULD CLEAN' : 'CLEANED'}: ${result.path}`)
    }
    console.log(`\nTotal files ${parsed.dryRun ? 'that would be cleaned' : 'cleaned'}: ${total}`)
    return 0
  } catch (error) {
    console.error(`cleanup-duplicate-docstrings: ${errorMessage(error)}`)
    return 1
  }
}

if (import.meta.main) {
  process.exitCode = await main()
}
